web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/CdsProfileFetcher.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.web.internal.correlation;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import org.apache.http.HttpResponse;
+import org.apache.http.ParseException;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
+import org.apache.http.impl.nio.client.HttpAsyncClients;
+import org.apache.http.nio.client.HttpAsyncClient;
+import org.apache.http.util.EntityUtils;
+
+public class CdsProfileFetcher implements AppProfileFetcher {
+
+	private HttpAsyncClient httpClient;
+    private String endpointAddress;
+    private static final String ProfileQueryEndpointAppIdFormat = "%s/api/profiles/%s/appId";
+    private static final String DefaultProfileQueryEndpointAddress = "https://dc.services.visualstudio.com";
+
+    // cache of tasks per ikey
+    private final ConcurrentHashMap<String, Future<HttpResponse>> tasks;
+
+    public CdsProfileFetcher() {
+        RequestConfig requestConfig = RequestConfig.custom()
```

Comment:
Ikey lookups need to be fast, otherwise we risk exhausting the hosting app's resources in high scale scenarios. If breeze takes more than a few seconds for ikey lookup, we already have a bigger problem. We might want to add a synchronous lookup during module init though, which i think is acceptable.